### ISSUE TITLE
Fix S6966 FP: Raises when methods have different signatures with optional parametes

### ISSUE THREAD

**Issue by rjgotten** [SonarSource/sonar-dotnet#9613]
### Description

S6966 conflates different synchronous and asynchronous signatures of a commonly named method - e.g. `DoWork` and `DoWorkAsync` - when both signatures use optional parameters and the caller does not pass parameters.

### Repro steps

```cs
file class Test
{
  public async Task<int> Run()
  {
    // falsely raises S6966 even though parameters are incompatible and on
    // actual real-world code may feature wildly different workloads.
    return DoWork();
  }

  public int DoWork(bool something = true) => 1;
  public async Task<int> DoWorkAsync(string somethingElse = "") => Task.CompletedTask(1);
}
```

### Expected behavior

S6966 is not raised, because the parameters are incompatible.

### Actual behavior

S6966 is raised, because it does not consider the difference in actual signature parameters and falsely believes the two calls are substitutable.

### Known workarounds

None. Suppress the broken rule.

### Related information

* C#/VB.NET Plugins version: whatever ships with SonarLint 8.2.0.96766
* Visual Studio version: 17.10.5
* MSBuild / dotnet version: dotnet SDK 8.0.303
* SonarScanner for .NET version (if used): N/A
* Operating System: Windows 11

**Comment by mary-georgiou-sonarsource** [comment]
Hey @rjgotten I confirm this is a false positive.
I've added it to our backlog.

**Comment by mary-georgiou-sonarsource** [comment]
Hello @rjgotten reseeing this it's not really a False positive.
Those two methods are not supposed to have the same signature. This is a not-so-good design to have - one of them should be named differently.

In case you cannot change the method names, you could accept the issue.

**Comment by martin-strecker-sonarsource** [comment]
Hello @rjgotten,

I'm not rely convinced by the example. Consider this factory  
```cs
public class Factory<T>
{
  public T Create(int key = 0) { }
  public Task<T> CreateAsync(string key = "") { } 
}
```
As a consumer of this API, I would think about such an API like so:

```cs
// In an async context
public async Task Consumer(Factory<Foo> fooFactory)
{
    var foo = fooFactory.Create(); // This is potentially blocking.
    var foo = await fooFactory.CreateAsync(); // This is the better "overload" as it isn't blocking.
}
```
The consumer can switch between both calls without changing the arguments of the invocation. How is the consumer supposed to know when it is okay to do the blocking call and when he should `await`? Both cases are indistinguishable from the consumer's point of view. Therefore, the recommendation to await the Task returning version inside the `async` method seems reasonable to me.

If the parameter isn't optional, the story is different, and in such cases, we do not raise an issue. We, in fact, re-bind with the async version of the method and let the compiler figure out if there is a matching overload that binds successfully without changing any of the arguments. Only if this succeeds we raise the issue.

Can you be more specific about the optional parameter? How does that parameter influence the method's asyncness? I have a hard time understanding how the `bool` versus a `string` from your first example can influence the method's asyncness.
